particle-iot/device-os
# Feature/gen2 otp store

### Problem

We'd like to store full 15-digit serial numbers when manufacturing Gen 2 devices in the OTP area, ideally using the same format used on Gen 3 devices. This requires support on the DeviceOS side as well.

### Solution

Implement `FLASH_ReadOTP()` and `FLASH_WriteOTP()` to read and write OTP on Gen 2 devices.

### Steps to Test
1. Compile and download below application
2. Observe the serial log, it will print serial number, mobile secret, and NCP ID


### Example App

```c
#include "deviceid_hal.h"
#include "platform_ncp.h"

SYSTEM_MODE(MANUAL);
Serial1LogHandler logHandler(115200, LOG_LEVEL_ALL, {
    {"app", LOG_LEVEL_ALL},
    {"hal.ble", LOG_LEVEL_WARN},
    {"system.listen.ble", LOG_LEVEL_WARN},
});

const size_t OTP_DEVICE_SERIAL_ADDRESS = 0;
const size_t OTP_DEVICE_SERIAL_SIZE = HAL_DEVICE_SERIAL_NUMBER_SIZE;
const size_t OTP_DEVICE_SECRET_ADDRESS = 16;
const size_t OTP_DEVICE_SECRET_SIZE = HAL_DEVICE_SECRET_SIZE;
const size_t OTP_NCP_ID_ADDRESS = 32;
const size_t OTP_NCP_ID_SIZE = 4;

extern "C" {
void __flash_acquire() {
    os_thread_scheduling(false, nullptr);
}

void __flash_release() {
    os_thread_scheduling(true, nullptr);
}
}

void setup() {

}

void loop() {
    uint8_t serialNumber[16] = {0};
    uint8_t mobileSecret[16] = {0};
    const uint8_t fixedSerialNumber[16] = {0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x3A, 0x3B, 0x3C, 0x3D, 0x3E, 0x00};
    const uint8_t fixedMobileSecret[16] = {0x60, 0x61, 0x62, 0x63, 0x64, 0x65, 0x66, 0x67, 0x68, 0x69, 0x6A, 0x6B, 0x6C, 0x6D, 0x6E, 0x00};
    int err_code;

    err_code = FLASH_WriteOTP(OTP_DEVICE_SERIAL_ADDRESS, fixedSerialNumber, sizeof(fixedSerialNumber));
    if (err_code == SYSTEM_ERROR_NOT_ALLOWED) {
        Log.info("Serial Number is initialized already");
    } else {
        Log.info("ERROR: %d", err_code);
    }
    err_code = FLASH_ReadOTP(OTP_DEVICE_SERIAL_ADDRESS, serialNumber, sizeof(serialNumber));
    if (err_code) {
        Log.info("ERROR: %d", err_code);
    }
    Log.info("Serial Number: %s", serialNumber);

    err_code = FLASH_WriteOTP(OTP_DEVICE_SECRET_ADDRESS, fixedMobileSecret, sizeof(fixedMobileSecret));
    if (err_code == SYSTEM_ERROR_NOT_ALLOWED) {
        Log.info("Mobile Secret is initialized already");
    } else {
        Log.info("ERROR: %d", err_code);
    }
    err_code = FLASH_ReadOTP(OTP_DEVICE_SECRET_ADDRESS, mobileSecret, sizeof(mobileSecret));
    if (err_code) {
        Log.info("ERROR: %d", err_code);
    }
    Log.info("Mobile Secret: %s", mobileSecret);

    const uint8_t fixedNcpId = MESH_NCP_BROADCOM_BCM9WCDUSI09;
    uint8_t ncpId = 0;
    err_code = FLASH_WriteOTP(OTP_NCP_ID_ADDRESS, &fixedNcpId, sizeof(fixedNcpId));
    if (err_code == SYSTEM_ERROR_NOT_ALLOWED) {
        Log.info("NCP is initialized already");
    } else {
        Log.info("ERROR: %d", err_code);
    }
    err_code = FLASH_ReadOTP(OTP_NCP_ID_ADDRESS, &ncpId, sizeof(ncpId));
    if (err_code) {
        Log.info("ERROR: %d", err_code);
    }
    Log.info("NCP ID: 0x%x", ncpId);

// Test this part with monolithic firmware
#if 0
    char sn[16] = {0};
    char mn[16] = {0};
    hal_get_device_serial_number(sn, sizeof(sn), nullptr);
    Log.info("hal_get_device_serial_number: %s", sn);
    hal_get_device_secret(mn, sizeof(mn), nullptr);
    Log.info("hal_get_device_secret: %s", mn);
#endif

    delay(1000);
}
```

### References

[CH38672]

---

### Completeness
